### ISSUE TITLE
How to handle O(n) modification methods on `list`

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#21933]
### Problem

The `list` type has a few methods that can add or remove elements from any index in the list. These operations cause its elements to be shifted in memory:

- `insert`: add one or more element at a given index - shifts rightward elements to the right
- `remove`: remove one or more elements matching a particular value - shifts rightward elements to the left
- `pop` (with index argument): removes and returns a value at a given index - shifts rightward elements to the left

As such, each can invalidate references into the list. Example:
```chapel
var l = new list(1..10);
ref x = l[5];
l.insert(2, 55); // x is no longer valid
```
Additionally, they may present challenges in implementing a distributed version of `list` with the same API.  Specifically, some distributed list implementations would require the entire data structure to be locked for end-only operations like `append`. This is because appending to the end of the list could, for example, race with an insertion earlier in the list that shifts elements into the same memory slot where the new element is being appended. 

Requiring an entire distributed list to be locked for operations on the end of the list would be a significant performance problem. However, if `insert`, `remove`, and `pop(i)` were not available on the distributed list, locking could be less aggressive.

### Proposal

Use a `param` argument in the list constructor to toggle the compile-time availability of: `insert`, `remove`, and `pop(i)`. 

This might look like:
```chapel
var l = new list(1..1000, endOnly=true);

l.insert(10, 55); // fails to compile
l.append(111); // okay
```

In the distributed list context, when `endOnly=true`, methods like `append` would be able to lock less aggressively. For example, with some implementations, only the locale the owns the end of the list would need to lock. 

---
*note: `sort` and `clear` could also be involved in this discussion. Since they modify the entire list, and invalidate references, they could also be excluded in end-only mode.

**Comment by jeremiah-corrado** [comment]
## Final Design Decision:

During a recent design meeting, a sub-team consisting of myself, @lydia-duncan, and @daviditen (with honorary member @mppf), decided to mark the following methods as unstable with the intention to deprecate:

- `list.insert`
- `list.remove`
- `list.pop` (overload that takes an index argument)

## Summary of Problem:

Our primary concern with continuing to support `insert`, `remove` and `pop` is two fold: 
1. in a serial context, they are inefficient, as they require elements to be shifted in memory (i.e., they are O(n) operations). 
2. in a parallel context, operations that modify the values in the list or modify the end of the list must lock against elements being shifted in memory. This can introduce an additional performance hit.

In practice, `insert`, `remove`, and `pop` are often used on small arrays so performance is not important; however, as we look ahead to implementing a distributed list with the same API as `list`, this concern becomes more significant — see the issue summary above for more.

## Discussion and Rationale:

During the design discussion, the following solutions for this problem were considered

1. **Status-quo:** use documentation to warn users about performance pitfalls but leave the API unchanged
    - Pro: doesn't disturb codes currently using these methods
    - Con: This may prevent us from being able to write a performant distributed list
2. create an **"end only" mode** on `list` that deactivates these methods
    - Pro: allows us to keep the convenience methods available on `list` and have a more efficient distributed list implementation when `endOnly=true` 
    - Con: requires us to maintain two distributed list implementations. May be difficult to inform users that they should turn `endOnly` on for performance critical codes 
3. make the methods **return a copy of the list** so that the O(n) nature of the operation is obvious
    - Pro: it will be easy to recognize (at an API level) that changing the size of a list in this manner is inefficient, and that a different algorithm/pattern should be used if one of these methods feels needed
    - Con: why support these operations if we intent do discourage their use so heavily?
4. **mark them as unstable** (intending to deprecate) and support them on a linked-list data structure in the standard library at some point in the future
    - Pro: will make it easier to write an efficient distributed list (with a single implementation)
    - Con: will be unstable until a linked-list (or linked-list like) data structure is added to the library

Overall, we decided to go with option 4. Removing `insert`, `remove` and `pop(idx)` from the `list` API will allow us to write better distributed/parallel versions of list. We do recognize the value of having methods like this available on a list data structure for convenience; however, we think it would be best to support those on a future `linkedList` collection type in the standard library rather than potentially compromising the performance and simplicity of a distributed list type.

<details>
<summary>voting breakdown</summary>

1. status quo: 1 🤔, 1 😕
2. end only mode: 2 👍, 1 🤔
3. copying versions: 2 🤔, 1 👍
4. mark unstable (to be deprecated): 4 👍

</details>

**Comment by dlongnecke-cray** [comment]
I have some concern over removing `insert`, `remove`, `pop`.

I don't think we should remove an operation from the API just because it will be slower or more difficult to implement in a distributed setting. In the worst case, can't we just document that such methods are slow, or warn/recommend against them? They are already understood to be slow in a local setting, so that they would be even more slow in a distributed setting is not much of a leap.

I get that if you're frequently adding/removing interior elements, then perhaps a resizable array is not the ADT of choice. But if we remove those methods, users don't really have any other recourse in the event if they do need to insert/remove an interior element.

As well, a user may be in a situation where they are using a `list` and cannot afford to switch to using a linked list, either because the memory cost is too expensive or their algorithm needs an array.

**Comment by jeremiah-corrado** [comment]
>  In the worst case, can't we just document that such methods are slow, or warn/recommend against them?

I'm not particularly worried about `myDistList.insert(1000, 5)` being slow. If that was the only issue, then I'd agree that documentation would be the solution.  I'm primarily worried about the effect they'll have on the overall implementation. Specifically, I think other operations on the list will need to lock the entire data structure (or do some kind of buffering of communication between locales) because values can be shifted around in memory at any time. Alternatively, if `insert`, `remove`, and `pop(idx)` aren't supported, then `myDistList[500] += 1`, for example, only needs to lock whichever locale that owns the 500th element.

**Comment by dlongnecke-cray** [comment]
I agree that it seems challenging to support these from an implementation perspective, but I think it would be a shame if even the serial, task-unsafe list did not have these methods. They seem like such a common convenience that, as a user, I think I would balk if I found out that they didn't exist.

I understand that we want to try and have the API remain uniform between serial/parallel/distributed variants of a data structure. I personally am OK with the idea of the performance of these operations being prohibitively slow on a distributed list. I would rather that than no methods at all.

At this point it feels like the only thing separating our `list` from a `stack` is that you can index into it.

**Comment by dlongnecke-cray** [comment]
Hmmm. You've obviously invested far more time into thinking about this than I have, so if it's going to be nigh on impossible from an implementation perspective then I'd downgrade my challenge to:

Maybe we should stop trying to have the serial/parallel/distributed variants of a data structure be perfectly consistent. We have interfaces on the way and we can have a "list" interface that expresses the minimal ADT in cases where that matters. I think that would be a better practice than expecting things to always work when writing unconstrained generics, anyway.

**Comment by jeremiah-corrado** [comment]
> Maybe we should stop trying to have the serial/parallel/distributed variants of a data structure be perfectly consistent. We have interfaces on the way and we can have a "list" interface that expresses the minimal ADT in cases where that matters. I think that would be a better practice than expecting things to always work when writing unconstrained generics, anyway.

I like this idea, but I worry that it would prevent us from stabilizing `list` if interfaces won't be ready in time for Chapel 2.0. 

Are you thinking that the distributed list could be an implementation of a separate interface (that wouldn't support the methods we've been discussing) which the standard `list` would also implement? And then we could get away with not creating a distributed list that implements entire "list" interface? In other words, we'd swap our current goal of: "make all the lists have the same interface" with "make all the lists implement the same 'vector' (or whatever it would be called) interface".

To respond to some of your earlier comments:

> But if we remove those methods, users don't really have any other recourse in the event that they do need to insert/remove an
> a user may be in a situation where they are using a list and cannot afford to switch to using a linked list

here is a rough picture of the future world I'm imaging that will hopefully clarify my position a bit.  We would eventually have two list data structures in the standard library:
1. a vector-like data structure with a pretty minimal interface that also supports an efficient distributed/parallel version
    - this is what I imagine the existing `list` will become
2. a generally less performant, but more convenient, list which may or may not be implemented as a linked-list that will be 
recommended for less performant critical tasks that would benefit from the expanded API
    - maybe a distributed/serial version could be implemented at some point, but would not be strictly needed

**Comment by dlongnecke-cray** [comment]
> separate interface (that wouldn't support the methods we've been discussing) which the standard list would also implement?

Yes, we would have some "ListInterface" that would be the most minimal ADT, and then if users needed a guarantee about a consistent interface across serial/parallel/distributed variants, they'd get it by using the interface type. Then their generic code will never break.

(As an aside, it feels like there are so many issues in 2.0 that it feels like interfaces would solve elegantly, that we have to jump through hoops to solve without them.)

> a generally less performant, but more convenient, list which may or may not be implemented as a linked-list that will be
recommended for less performant critical tasks that would benefit from the expanded API

I worry that we are sacrificing too much simplicity/convenience to fulfill the goal of plug-&-play across serial/parallel/distributed versions in unconstrained generic code.

To me having both a `fastList` and a `flexibleList` seems like overkill.

If `flexibleList` is basically just a linked list, then I'd argue we're back to square one, because there are situations where I'd like to insert/remove from the interior of a vector but still get the O(1) indexing guarantee. And if `list` doesn't have those methods, I have to go and build my own collection that does.

If it's not, then I'd argue we've created unnecessary maintenance and documentation burden for ourselves. Is this new type also a type with 3 variants? Then we've still got the implementation difficulty you described. If not, then we created a new type just to replace a serial list. Feels like we should just add methods to the serial list.

**Comment by bradcray** [comment]
What makes the list a list rather than a stack if it doesn't have insert and remove calls?  Is it simply the ability to iterate over values and inspect intermediate ones rather than only having the top-of-stack be visible?

**Comment by jeremiah-corrado** [comment]
> What makes the list a list rather than a stack if it doesn't have insert and remove calls? Is it simply the ability to iterate over values and inspect intermediate ones rather than only having the top-of-stack be visible?

Yes and there is also the ability to modify intermediate values with `set`, `update`, or `this()`.

**Comment by bradcray** [comment]
To me, option 4 seems sufficiently unlike what I think of a list as being that if we were to proceed with it, I think we should rename the collection to something else.  If you were to ask me what operations a typical list supports, insert and remove come to mind right near the top of the list (and far before "modify-element-in-place" does).  To me, the restricted interface feels more like a stack with some nice/extra convenience features (and a few that scare me) than a list.  Is there a precedent in any of our favorite languages for supporting a data structure named `list` that supports the subset of operations in option 4?

Have we surveyed existing uses of `list` in Chapel to see whether current uses will stop working if/when these methods are removed?  (this is more curiosity than anything b/c `list` hasn't existed for so long that I imagine there being tons of uses of it).

> Specifically, I think other operations on the list will need to lock the entire data structure (or do some kind of buffering of communication between locales) because values can be shifted around in memory at any time.

It's hard for me to know how bad something like a global lock would be vs. not—I haven't come across real-world applications that want to use distributed lists to know what they'd expect performance-wise (or whether a list is even what they want — versus, say, a distributed bag or the like).  Have others?

**Comment by mppf** [comment]
> To me, option 4 seems sufficiently unlike what I think of a list as being that if we were to proceed with it, I think we should rename the collection to something else. If you were to ask me what operations a typical list supports, insert and remove come to mind right near the top of the list (and far before "modify-element-in-place" does).

For those with CS training, I don't think it's very far from these reactions to "And insert in the middle should be O(1)". So maybe we just need to rename it. But, even with a linked list, inserting at index `n` is not O(1) because it is O(n) work to find the list node for index `n`. Which indicates to me that the API doesn't make sense for (efficient) inserting in the middle anyway.

> I haven't come across real-world applications that want to use distributed lists to know what they'd expect performance-wise

One use case is to have basically a distribute work queue. See also #21958 and the related efforts in DistBag / DistQueue with the test CollectionNQueens (which can use either data structure, but that's more of a test than a motivation for DistQueue). But for a distributed work queue, "list" semantics/API is probably more restrictive of the implementation than necessary.

Anyway, if we don't feel confident that we understand what a distributed list is or what its use cases are, I don't think we should be holding ourselves to the idea that the distributed list must be a strict superset of the API of the serial list and that therefore the serial list design must be constrained by the distributed list.

**Comment by jeremiah-corrado** [comment]
> Have we surveyed existing uses of list in Chapel to see whether current uses will stop working if/when these methods are removed? (this is more curiosity than anything b/c list hasn't existed for so long that I imagine there being tons of uses of it).

In the testing system, I only found a few cases that use `insert` other than the tests that are specifically for it.

> Anyway, if we don't feel confident that we understand what a distributed list is or what its use cases are, I don't think we should be holding ourselves to the idea that the distributed list must be a strict superset of the API of the serial list and that therefore the serial list design must be constrained by the distributed list.

I'm okay with this as well. Overall, my initial thinking with proposal 4 was that `list`'s API would be focused on performance in distributed and parallel settings, and that we'd eventually add another data structure which is more focused on convenience (and has a more typical list API). However, given the above discussion and some offline discussions I think it makes sense to do the exact opposite. I.e., keep `list`'s API generally as is, and eventually create another list-like data structure whose API is more geared towards a parallel/distributed implementation. With this approach we could still create a distributed version of `list` with the same API while making it clear that another data structure, like a distributed queue, might offer better performance.

**Comment by bradcray** [comment]
> Anyway, if we don't feel confident that we understand what a distributed list is or what its use cases are, I don't think we should be holding ourselves to the idea that the distributed list must be a strict superset of the API of the serial list and that therefore the serial list design must be constrained by the distributed list.

> and eventually create another list-like data structure whose API is more geared towards a parallel/distributed implementation.

Recall that we recently received user feedback that the parallel list should be the basis for the common list API and that it would be OK for the serial list to support routines that the parallel list did not (or at least, that's how I interpreted the comment).  I don't think we've agreed to that philosophy internally (so the quoted comments here arguably still reflect the previous philosophy which is arguably the current one), but I found the general argument compelling.

**Comment by jeremiah-corrado** [comment]
> it would be OK for the serial list to support routines that the parallel list did not 

When we were having this discussion offline, I was working under the assumption that the APIs needed to be identical (modulo additional initializer arguments), but this approach certainly simplifies this discussion and makes proposal 1 seem the most favorable to me. I.e., I like the idea of keeping the `remove`, `insert`, and `pop(i)` methods on the serial `list` and creating a distributed list that does not include them.

**Comment by jeremiah-corrado** [comment]
After some further discussion offline with the ad-hoc subteam (@lydia-duncan, @daviditen, and I), as well as a few others, we decided not to deprecate `list.insert`, `list.remove` and `list.pop(index)`.

**Distributed list discussion:**

The initial reasoning for the deprecation was that we wanted `list`'s API to be designed for an optimal distributed list implementation. This meant essentially making `list` be an "indexable stack" whose size can only be modified from the end of the list (i.e., we were prioritizing the distributed list's performance over the serial lists usability/convenience). This proposal also came along with the idea that we'd add something like a serial linked-list to the standard library so that the methods discussed in this issue would still be available to users in some form. 

After more discussion, we've decided to go with the inverse of the initial proposal. Instead of relying on another serial data structure to support these "convenience" methods, we'll rely on other distributed data structures to support the high performance distributed case. In other words, `list` will stay on the convenience side of the performance/convenience tradeoff — keeping `insert`, `pop`, and `remove`, and other distributed data structures like a `stack` and/or `queue` can eventually be created to sit on the performance side of that tradeoff.

I'll make a note of this proposed future direction in the documentation for the distributed list prototype.

**Amended Final Decision (as it relates to 2.0 stabilization):**

`insert` and `remove` will be stabilized as-is, and `pop(idx)` will be renamed to `getAndRemove`.

This overload of `pop` is being renamed because most of us found it to be counter intuitive that a "pop" method could refer to elements not at the end of a list. The name `getAndRemove` was selected because it mirrors the analogous method on `map` and clearly describes what the method does. We also considered making it a second overload of `remove`; however, there would be a disambiguation issue with list's of integers (i.e., is the user trying to remove all the `5`'s from the list, or are they trying to remove the 5th element?)
